math/src/field/f64/mod.rs

```diff
@@ -53,26 +59,95 @@ const ELEMENT_BYTES: usize = core::mem::size_of::<u64>();
 /// Internal values are stored in the range [0, 2^64). The backing type is `u64`.
 #[derive(Copy, Clone, Debug, Default)]
 pub struct BaseElement(u64);
-
 impl BaseElement {
-    /// Creates a new field element from the provided `value`. If the value is greater than or
-    /// equal to the field modulus, modular reduction is silently performed.
-    pub const fn new(value: u64) -> Self {
-        Self(value % M)
+    /// Creates a new field element from the provided `value`; the value is converted into
+    /// Montgomery representation.
+    pub const fn new(value: u64) -> BaseElement {
+        BaseElement(BaseElement::mont_red((value as u128) * (R2 as u128)))
+    }
+    /// Gets the inner value that might not be canonical
+
+    pub const fn inner(self: &Self) -> u64 {
+        return self.0;
+    }
+
+    /// Montgomery reduction
+    #[inline(always)]
+    pub const fn mont_red(x: u128) -> u64 {
+        // See reference above for a description of the following implementation.
+        let xl = x as u64;
+        let xh = (x >> 64) as u64;
+        let (a, e) = xl.overflowing_add(xl << 32);
+
+        let b = a.wrapping_sub(a >> 32).wrapping_sub(e as u64);
+
+        let (r, c) = xh.overflowing_sub(b);
+        r.wrapping_sub(0u32.wrapping_sub(c as u32) as u64)
+    }
```

Comment:
I'd move this function into the helper functions (where `mont_reduce()` is now). We should also probably give them distinct enough names to make it difficult to confuse them. For example, `mod_reduce_ct()` (for constant time) and `mont_redcue_vt()` (for variable time).

Reply:
Should we keep both or should we call both mont_reduce() and comment one out? This way it would be easier to test the overall effect of constant vs. variable time implementations.

Reply:
I think we should keep both - just have suffixes to clear differentiate them. Above I suggested `mont_reduce_ct()` and `mont_reduce_vt()` - but other names could work too.